File: general-solving/leetcode/0151_reverse-words-in-a-string.cs
/***************************************************************************************************
* Title : Reverse Words in a String
* URL   : https://leetcode.com/problems/reverse-words-in-a-string/
* Date  : 2015-09-29
* Author: Atiq Rahman
* Status: Accepted
* Comp  : O(n) considering Array.Reverse (confirm)
* Notes : Look after source code for detail analysis of the solution
*   Bloomberg LP has an interesting variation I think.
* meta  : tag-string, tag-company-bloomberg, tag-leetcode-medium
***************************************************************************************************/
using System.Linq;

public class Solution
{
  /// <summary>
  /// Simply returning
  ///  string.Join(' ', tokens).Trim()
  /// does not work since there are emtpy strings in tokens or contains whitespaces only
  /// Hence, utilize linq to weed out empty strings
  /// </summary>
  /// <param name="s"></param>
  /// <returns> word reversed string with unnecessary spaces removed </returns>
  public string ReverseWords(string s) {
    string[] tokens= s.Split(' ');
    Array.Reverse(tokens);
    return string.Join(' ', tokens.Where(token => !string.IsNullOrEmpty(token)));
  }
}

/* First version,
  StringBuilder result = new StringBuilder();
    
  foreach (var str in tokens)
    if (str.Length != 0)
      result.Append(str+" ");

  return result.ToString().TrimEnd(' ');   
*/

File: general-solving/leetcode/0151_reverse-words-in-a-string.cpp
/***************************************************************************
* Title : Reverse Words in a String
* URL   : https://leetcode.com/problems/reverse-words-in-a-string
* Date  : 2018-02-20
* Author: Atiq Rahman
* Comp  : Time O(N), Space O(N)
* Status: Accepted
* Notes : using a new string to store the reversed result
*   a prefered solution is to use O(1) space probably using C programing
*   language
*   As we are allowed to modify passed string we can reverse the whole string
*   in one iteration. In another iteration, we can reverse the words inside it
*   got get the expected result
* meta  : tag-string, tag-company-bloomberg
***************************************************************************/
#include <sstream>
#include <string>
#include <vector>

class Solution {
private:
  std::string trim(const std::string& str, const std::string& whitespace =
    " \t") {
    const auto strBegin = str.find_first_not_of(whitespace);
    if (strBegin == std::string::npos)
      return ""; // no content

    const auto strEnd = str.find_last_not_of(whitespace);
    const auto strRange = strEnd - strBegin + 1;

    return str.substr(strBegin, strRange);
  }

public:
  void reverseWords(string &str) {
    std::istringstream ss(str);
    std::string token, rev;
    char sep = ' ';
    while (std::getline(ss, token, sep))
      if (token.empty() == false && token != " ")
        rev = token + " " + rev;
    str = trim(rev);
  }
};
